### PR TITLE
FEC-12761 IMA SDK iOS + tvOS updates 3.18.1

### PR DESCRIPTION
**Pull request by chausov**
Solves 
FEC-12761

### CODE REVIEW

File: Sources/IMADAIPlugin.swift [hunk 1]
```diff
@@ -196,7 +196,10 @@ import PlayKitUtils
             throw IMADAIPluginRequestError.missingPlayerView
         }
         
-        adDisplayContainer = IMADAIPlugin.createAdDisplayContainer(forView: playerView, viewController: playerView.findViewController(), withCompanionView: pluginConfig.companionView)
+        let displayContainer = IMADAIPlugin.createAdDisplayContainer(forView: playerView,
+                                                                     viewController: playerView.findViewController(),
+                                                                     withCompanionView: pluginConfig.companionView)
+        adDisplayContainer = displayContainer
```
Comment (giladna) [inline Sources/IMADAIPlugin.swift:202]: @chausov  why not to use it as it is today?

File: Sources/IMADAIPlugin.swift [hunk 2]
```diff
@@ -404,33 +407,28 @@ import PlayKitUtils
...
         
-        guard let adError = adErrorData.adError else {
-            PKLog.error("AdsLoader faild with error.")
-            return
-        }
-        let adErrorMessage: String = adError.message == nil ? "" : adError.message
-        PKLog.error(adErrorMessage)
-        messageBus?.post(AdEvent.Error(nsError: IMAPluginError(adError: adError).asNSError))
-        delegate?.adsPlugin(self, loaderFailedWith: adErrorMessage)
+        PKLog.error(adErrorData.adError.message)
```
Comment (giladna) [inline Sources/IMADAIPlugin.swift]: @chausov that mathod may have error or not why did you remove the validation if there is error or not?

File: Sources/IMADAIPlugin.swift [hunk 3]
```diff
@@ -455,39 +453,60 @@ import PlayKitUtils
                 }
             }
         case .LOADED:
-            let adEvent = event.ad != nil ? AdEvent.AdLoaded(adInfo: PKAdInfo(ad: event.ad,
-                                                                              podCount: nil,
-                                                                              adPlayHead: .nan)) : AdEvent.AdLoaded()
+            var adEvent = AdEvent.AdLoaded()
```
Comment (giladna) [inline Sources/IMADAIPlugin.swift:458]: @chausov  I think we can move it to a method

File: Sources/IMAPlugin.swift [hunk 4]
```diff
@@ -421,14 +429,17 @@ enum IMAState: Int, StateProtocol, CustomStringConvertible {
                 self.discardAdBreak(adsManager: adsManager)
             } else {
                 var adEvent = AdEvent.AdLoaded()
-                if event.ad != nil {
-                    let adInfo = PKAdInfo(ad: event.ad,
-                                          podCount: adsManager?.adCuePoints.count,
+                
+                if let ad = event.ad {
```
Comment (giladna) [inline Sources/IMAPlugin.swift:435]: exctract to method

File: Sources/IMAPlugin.swift [hunk 5]
```diff
@@ -228,29 +228,38 @@ enum IMAState: Int, StateProtocol, CustomStringConvertible {
             throw IMAPluginRequestError.emptyAdTag
         }
         
-        adDisplayContainer = IMAPlugin.createAdDisplayContainer(forView: playerView, viewController: playerView.findViewController(), withCompanionView: self.config.companionView)
+        let displayContainer = IMAPlugin.createAdDisplayContainer(forView: playerView,
+                                                                  viewController: playerView.findViewController(),
+                                                                  withCompanionView: self.config.companionView)
+        
+        adDisplayContainer = displayContainer
```
Comment (giladna) [inline Sources/IMAPlugin.swift:235]: is there a reason for that?

File: Sources/IMADAIPlugin.swift (hunk 1 above)
Comment (chausov) [reply Sources/IMADAIPlugin.swift:202]: `adDisplayContainer` defined as optional but in `IMALiveStreamRequest` it required non optional `adDisplayContainer:` as a parameter.
we may change the definition of `adDisplayContainer` to be non optional property it will be better.

This changes required because in IMA SDK some of the types was changed

File: Sources/IMADAIPlugin.swift (hunk 2 above)
Comment (chausov) [reply Sources/IMADAIPlugin.swift]: adErrorData now is non Optional type so `guard let adError = adErrorData.adError` does not make a sense

File: Sources/IMADAIPlugin.swift (hunk 3 above)
Comment (chausov) [reply Sources/IMADAIPlugin.swift:458]: we can not wrap it to a method, because in each case there are different `AdEvent` created
in total 6 different methods needed, it will not save any lines of code: 
```Swift
AdEvent.AdLoaded()
AdEvent.AdStarted()
AdEvent.AdPaused()
AdEvent.AdResumed()
AdEvent.AdClicked()
```

File: Sources/IMAPlugin.swift (hunk 5 above)
Comment (chausov) [reply Sources/IMAPlugin.swift:235]: Same reason as in `IMADAIPlugin.swift`


`adDisplayContainer` defined as optional but in `IMALiveStreamRequest` it required non optional `adDisplayContainer:` as a parameter.
we may change the definition of `adDisplayContainer` to be non optional property it will be better.

This changes required because in IMA SDK some of the types was changed

File: Sources/IMAPlugin.swift (hunk 4 above)
Comment (chausov) [reply Sources/IMAPlugin.swift:435]: Same problem like in `IMADAIPlugin.swift`

**Review by giladna — APPROVED** [review]
